File: app/src/main/java/org/mifos/mobilebanking/ui/activities/base/BaseActivity.java
package org.mifos.mobilebanking.ui.activities.base;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import com.mifos.mobile.passcode.BasePassCodeActivity;

import org.mifos.mobilebanking.MifosSelfServiceApp;
import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.injection.component.ActivityComponent;
import org.mifos.mobilebanking.injection.component.DaggerActivityComponent;
import org.mifos.mobilebanking.injection.module.ActivityModule;
import org.mifos.mobilebanking.ui.activities.PassCodeActivity;
import org.mifos.mobilebanking.ui.views.BaseActivityCallback;
import org.mifos.mobilebanking.utils.LanguageHelper;

/**
 * @author ishan
 * @since 08/07/16
 */
@SuppressLint("Registered")
public class BaseActivity extends BasePassCodeActivity implements BaseActivityCallback {

    protected Toolbar toolbar;
    private ActivityComponent activityComponent;
    private ProgressDialog progress;

    @Override
    public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);
        toolbar = findViewById(R.id.toolbar);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            setToolbarElevation();
        }
    }

    /**
     * Used for removing elevation from toolbar
     */
    public void hideToolbarElevation() {
        ViewCompat.setElevation(toolbar, 0);
    }

    /**
     * Used for setting toolbar elevation
     */
    public void setToolbarElevation() {
        ViewCompat.setElevation(toolbar, 8f);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    /**
     * Used for dependency injection
     * @return {@link ActivityComponent} which is used for injection
     */
    public ActivityComponent getActivityComponent() {
        if (activityComponent == null) {
            activityComponent = DaggerActivityComponent.builder()
                    .activityModule(new ActivityModule(this))
                    .applicationComponent(MifosSelfServiceApp.get(this).component())
                    .build();
        }
        return activityComponent;
    }

    /**
     * This method is use to provide back button feature in the toolbar of activities
     */
    protected void showBackButton() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Displays a toast in current activity. In this method the duration
     * supplied is Short by default. If you want to specify duration
     * use {@link BaseActivity#showToast(String, int)} method.
     *
     * @param message Message that the toast must show.
     */
    public void showToast(String message) {
        showToast(message, Toast.LENGTH_SHORT);
    }

    /**
     * Displays a toast in current activity. The duration can of two types:
     * <ul>
     * <li>SHORT</li>
     * <li>LONG</li>
     * </ul>
     *
     * @param message   Message that the toast must show.
     * @param toastType Duration for which the toast must be visible.
     */
    public void showToast(@NonNull String message, @NonNull int toastType) {
        Toast.makeText(BaseActivity.this, message, toastType).show();
    }

    /**
     * Calls a method {@code showProgressDialog("Working")} which displays ProgressDialog
     */
    public void showProgressDialog() {
        showProgressDialog(getString(R.string.working));
    }

    /**
     * Displays a ProgressDialog
     * @param message Message you want to display in Progress Dialog
     */
    @Override
    public void showProgressDialog(String message) {
        if (progress == null) {
            progress = new ProgressDialog(this, ProgressDialog.STYLE_SPINNER);
            progress.setCancelable(false);
        }
        progress.setMessage(message);
        progress.show();

    }

    /**
     * Hides the progress dialog if it is currently being shown
     */
    @Override
    public void hideProgressDialog() {
        if (progress != null && progress.isShowing()) {
            progress.dismiss();
            progress = null;
        }
    }

    /**
     * Used for setting title of Toolbar
     * @param title String you want to display as title
     */
    public void setActionBarTitle(String title) {
        if (getSupportActionBar() != null && getTitle() != null) {
            setTitle(title);
        }
    }

    protected void setActionBarTitle(int title) {
        setActionBarTitle(getResources().getString(title));
    }

    /**
     * @return Returns toolbar linked with current activity
     */
    public Toolbar getToolbar() {
        return toolbar;
    }

    /**
     * Calls {@code setActionBarTitle()} to set Toolbar title
     * @param title String you want to set as title
     */
    @Override
    public void setToolbarTitle(String title) {
        setActionBarTitle(title);
    }

    @Override
    public Class getPassCodeClass() {
        return PassCodeActivity.class;
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(LanguageHelper.onAttach(base));
    }

    /**
     * Replace Fragment in FrameLayout Container.
     *
     * @param fragment Fragment
     * @param addToBackStack Add to BackStack
     * @param containerId Container Id
     */
    public void replaceFragment(Fragment fragment, boolean addToBackStack, int containerId) {
        invalidateOptionsMenu();
        String backStateName = fragment.getClass().getName();
        boolean fragmentPopped = getSupportFragmentManager().popBackStackImmediate(backStateName,
                0);

        if (!fragmentPopped && getSupportFragmentManager().findFragmentByTag(backStateName) ==
                null) {
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            transaction.replace(containerId, fragment, backStateName);
            if (addToBackStack) {
                transaction.addToBackStack(backStateName);
            }
            transaction.commit();
        }
    }

    /**
     * It pops all the fragments which are currently in the backStack
     */
    public void clearFragmentBackStack() {
        FragmentManager fm = getSupportFragmentManager();
        int backStackCount = getSupportFragmentManager().getBackStackEntryCount();
        for (int i = 0; i < backStackCount; i++) {
            int backStackId = getSupportFragmentManager().getBackStackEntryAt(i).getId();
            fm.popBackStack(backStackId, FragmentManager.POP_BACK_STACK_INCLUSIVE);
        }
    }

    public int stackCount() {
        return getSupportFragmentManager().getBackStackEntryCount();
    }


    public static void hideKeyboard(Context context) {
        Activity activity = (Activity) context;
        InputMethodManager inputMethodManager =
                (InputMethodManager) context.getSystemService(
                        Activity.INPUT_METHOD_SERVICE);
        inputMethodManager.hideSoftInputFromWindow(
                activity.getCurrentFocus().getWindowToken(), 0);
    }
}

File: app/src/main/java/org/mifos/mobile/ui/activities/base/BaseActivity.kt
package org.mifos.mobile.ui.activities.base

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.os.Bundle
import android.view.MenuItem
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.appcompat.widget.Toolbar

import androidx.core.view.ViewCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager

import com.mifos.mobile.passcode.BasePassCodeActivity
import org.mifos.mobile.MifosSelfServiceApp
import org.mifos.mobile.R
import org.mifos.mobile.injection.component.ActivityComponent
import org.mifos.mobile.injection.component.DaggerActivityComponent
import org.mifos.mobile.injection.module.ActivityModule
import org.mifos.mobile.ui.activities.PassCodeActivity
import org.mifos.mobile.ui.views.BaseActivityCallback
import org.mifos.mobile.utils.LanguageHelper

/**
 * @author ishan
 * @since 08/07/16
 */
@SuppressLint("Registered")
open class BaseActivity : BasePassCodeActivity(), BaseActivityCallback {
    /**
     * @return Returns toolbar linked with current activity
     */
    var toolbar: Toolbar? = null
        protected set

    /**
     * Used for dependency injection
     * @return [ActivityComponent] which is used for injection
     */
    var activityComponent: ActivityComponent? = null
        get() {
            if (field == null) {
                field = DaggerActivityComponent.builder()
                        .activityModule(ActivityModule(this))
                        .applicationComponent(MifosSelfServiceApp.get(this).component())
                        .build()
            }
            return field
        }
        private set
    private var progress: ProgressDialog? = null
    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)
        toolbar = findViewById(R.id.toolbar)
        if (toolbar != null) {
            setSupportActionBar(toolbar)
            setToolbarElevation()
        }
    }

    /**
     * Used for removing elevation from toolbar
     */
    fun hideToolbarElevation() {
        ViewCompat.setElevation(toolbar!!, 0f)
    }

    /**
     * Used for setting toolbar elevation
     */
    fun setToolbarElevation() {
        ViewCompat.setElevation(toolbar!!, 8f)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    /**
     * This method is use to provide back button feature in the toolbar of activities
     */
    protected fun showBackButton() {
        if (supportActionBar != null) {
            supportActionBar!!.setHomeButtonEnabled(true)
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Displays a toast in current activity. In this method the duration
     * supplied is Short by default. If you want to specify duration
     * use [BaseActivity.showToast] method.
     *
     * @param message Message that the toast must show.
     */
    fun showToast(message: String) {
        showToast(message, Toast.LENGTH_SHORT)
    }

    /**
     * Displays a toast in current activity. The duration can of two types:
     *
     *  * SHORT
     *  * LONG
     *
     *
     * @param message   Message that the toast must show.
     * @param toastType Duration for which the toast must be visible.
     */
    fun showToast(message: String, toastType: Int) {
        Toast.makeText(this@BaseActivity, message, toastType).show()
    }

    /**
     * Calls a method `showProgressDialog("Working")` which displays ProgressDialog
     */
    fun showProgressDialog() {
        showProgressDialog(getString(R.string.working))
    }

    /**
     * Displays a ProgressDialog
     * @param message Message you want to display in Progress Dialog
     */
    override fun showProgressDialog(message: String?) {
        if (progress == null) {
            progress = ProgressDialog(this, ProgressDialog.STYLE_SPINNER)
            progress!!.setCancelable(false)
        }
        progress!!.setMessage(message)
        progress!!.show()
    }

    /**
     * Hides the progress dialog if it is currently being shown
     */
    override fun hideProgressDialog() {
        if (progress != null && progress!!.isShowing) {
            progress!!.dismiss()
            progress = null
        }
    }

    /**
     * Used for setting title of Toolbar
     * @param title String you want to display as title
     */
    fun setActionBarTitle(title: String?) {
        if (supportActionBar != null && getTitle() != null) {
            setTitle(title)
        }
    }

    protected fun setActionBarTitle(title: Int) {
        setActionBarTitle(resources.getString(title))
    }

    /**
     * Calls `setActionBarTitle()` to set Toolbar title
     * @param title String you want to set as title
     */
    override fun setToolbarTitle(title: String?) {
        setActionBarTitle(title)
    }

    override fun getPassCodeClass(): Class<*> {
        return PassCodeActivity::class.java
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(LanguageHelper.onAttach(base))
    }

    /**
     * Replace Fragment in FrameLayout Container.
     *
     * @param fragment Fragment
     * @param addToBackStack Add to BackStack
     * @param containerId Container Id
     */
    fun replaceFragment(fragment: Fragment, addToBackStack: Boolean, containerId: Int) {
        invalidateOptionsMenu()
        val backStateName = fragment.javaClass.name
        val fragmentPopped = supportFragmentManager.popBackStackImmediate(backStateName,
                0)
        if (!fragmentPopped && supportFragmentManager.findFragmentByTag(backStateName) ==
                null) {
            val transaction = supportFragmentManager.beginTransaction()
            transaction.replace(containerId, fragment, backStateName)
            if (addToBackStack) {
                transaction.addToBackStack(backStateName)
            }
            transaction.commit()
        }
    }

    /**
     * It pops all the fragments which are currently in the backStack
     */
    fun clearFragmentBackStack() {
        val fm = supportFragmentManager
        val backStackCount = supportFragmentManager.backStackEntryCount
        for (i in 0 until backStackCount) {
            val backStackId = supportFragmentManager.getBackStackEntryAt(i).id
            fm.popBackStack(backStackId, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        }
    }

    fun stackCount(): Int {
        return supportFragmentManager.backStackEntryCount
    }

    companion object {
        fun hideKeyboard(context: Context) {
            val activity = context as Activity
            val inputMethodManager = context.getSystemService(
                    Activity.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(
                    activity.currentFocus.windowToken, 0)
        }
    }
}